tests_e2e/tests/agent_removal/agent_removal.py

```diff
@@ -0,0 +1,118 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import json
+import os
+from time import sleep
+from assertpy import fail, assert_that
+
+from azurelinuxagent.common.protocol.extensions_goal_state_from_vm_settings import _CaseFoldedDict
+from azurelinuxagent.common.utils.archive import _EXT_CONF_FILE_NAME, _VM_SETTINGS_FILE_NAME
+from azurelinuxagent.common.utils.textutil import parse_doc, find, findall, findtext, gettext
+from tests_e2e.tests.lib.agent_test import AgentVmTest
+from tests_e2e.tests.lib.agent_test_context import AgentVmTestContext
+from tests_e2e.tests.lib.logging import log
+from tests_e2e.tests.lib.ssh_client import SshClient
+
+
+class AgentRemoval(AgentVmTest):
+    """
+    This test verifies the manifest contents after agent removal from PIR.
+    """
+
+    def __init__(self, context: AgentVmTestContext):
+        super().__init__(context)
+        self._ssh_client: SshClient = self._context.create_ssh_client()
+        self._expected_versions = None
+        self._removed_version = None
+        if hasattr(context, "expected_versions") and hasattr(context, "removed_version"):
+            fail("Only one of the following arguments should be provided, but both were: expected_versions, removed_version")
+        elif not hasattr(context, "expected_versions") and not hasattr(context, "removed_version"):
+            fail("Exactly one of the following arguments should be provided: expected_versions, removed_version")
+        elif hasattr(context, "expected_versions"):
+            self._expected_versions = context.expected_versions.split(';')
+        elif hasattr(context, "removed_version"):
+            self._removed_version = context.removed_version
+        self._GAFamily = context.GAFamily if hasattr(context, "GAFamily") else "Prod"
+
+    def run(self):
+        log.info("Retrieving agent manifests for {0} GAFamily...".format(self._GAFamily))
+
+        manifest_uris = self._get_manifest_uris()
+
+        log.info("")
+        log.info("Checking agent versions in manifest URIs from goal state...")
+        for uri in manifest_uris:
+            log.info("")
+            log.info("URI: {0}".format(uri))
+            agent_versions = self._ssh_client.run_command("curl -s {0} | grep -oP '(?<=<Version>).*?(?=</Version>)'".format(uri)).splitlines()
+            log.info("Agent versions in manifest: {0}".format(agent_versions))
+            if self._expected_versions is not None:
+                log.info("Expected versions: {0}".format(self._expected_versions))
+                assert_that(self._expected_versions).described_as("Expected agent versions in manifest does not match actual agent versions in manifest. Expected={0}, Actual={1}".format(self._expected_versions, agent_versions)).is_equal_to(agent_versions)
+                log.info("Expected versions match actual agent versions in manifest")
+            else:
```

Comment:
can we support expected versions and removed versions test in the same run?

Reply:
What is the advantage of supporting both in the same run? Expected versions list should not contain the versions we removed

Reply:
I'm with you on Expected vessions but I don't get why we need removed version test. Expected version helping all the cases right, i dont see when we need removed version check

Reply:
The versions may be different in canary than rest of Prod. If running on all regions, removed_version should be used if canary versions are different than rest of prod